src/9FC031D0.c

```diff
@@ -1,33 +1,234 @@
 #include "include_asm.h"
 #include "ultratypes.h"
+#include "bbtypes.h"
+#include "libcrypto/sha1.h"
+#include "PR/bcp.h"
 
-INCLUDE_ASM("asm/non_matchings/9FC031D0", virage2_gen_public_key);
+void* wordcopy(void* dst, void* src, s32 nWords);
+void* memcpy(void* dst, void* src, u32 num);
 
-INCLUDE_ASM("asm/non_matchings/9FC031D0", set_proc_permissions);
+extern Virage2* virage2_offset;
 
-INCLUDE_ASM("asm/non_matchings/9FC031D0", rsa_verify_signature);
+void virage2_gen_public_key(u32* pubkeyOut) {
+    u32 i;
 
-INCLUDE_ASM("asm/non_matchings/9FC031D0", rsa_check_signature);
+    for(i = 0; i < 0x10; i++) {
+        if (virage2_offset->publicKey[i] != 0) {
+            wordcopy(pubkeyOut, virage2_offset->publicKey, 0x10);
+            return;
+        }
+    }
+
+    eccGenPublicKey(pubkeyOut, virage2_offset->privateKey);
+    wordcopy(virage2_offset->publicKey, pubkeyOut, 0x10);
+}
+
+extern u32 cur_proc_allowed_skc_bitmask;
+
+void set_proc_permissions(BbContentMetaDataHead* cmdHead) {
+    s32 temp;
 
-INCLUDE_ASM("asm/non_matchings/9FC031D0", func_9FC03410);
+    IO_WRITE(PI_ALLOWED_IO, (u8)cmdHead->hwAccessRights);
```

Comment:
```suggestion
    IO_WRITE(PI_ALLOWED_IO, cmdHead->hwAccessRights & 0xFF);
```
Since the others are expressed in terms of masks and shifts, this could be more uniform